### PR TITLE
feat(aws): Refresh expired session tokens

### LINKED ISSUE #15: the expired session token is not handled gracefully
here is what I get (redacted):

```
E20220924 22:13:38.256793 1244660 s3.cc:143] http error: HTTP/1.1 400 Bad Request
x-amz-request-id: BLA
x-amz-id-2: FOOO
Content-Type: application/xml
Transfer-Encoding: chunked
Date: Sat, 24 Sep 2022 19:13:37 GMT
Server: AmazonS3
Connection: close

4ac
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>ExpiredToken</Code><Message>The provided token has expired.</Message><Token-0>TOKENFOOBAR</Token-0><RequestId>FOO</RequestId><HostId>BARR</HostId></Error>
0

F20220924 22:13:38.258005 1244659 s3_demo.cc:51] Check failed: list_res 
```

### PR DESCRIPTION
**Pull request by dranikpg**
Didn't test it yet on real EC2 (only on metadata mock), waiting for my token to expire 👀 

Fixes #15

### CODE REVIEW

File: util/cloud/aws.h [hunk 1]
```diff
@@ -3,46 +3,61 @@
...
-  // TODO: we should remove domain argument in favor to subdomain (bucket).
-  // and build the whole domain it from service and region
-  // for example, "<bucket>.s3.eu-west-1.amazonaws.com"
-  // See: https://docs.aws.amazon.com/general/latest/gr/s3.html
-  //
-  void Sign(std::string_view payload_sig, HttpHeader* header) const;
+  // Sign request and send it with client->Send() internally. Try to refresh session token if needed.
+  std::error_code SendRequest(std::string_view payload_sig, http::Client* client,
+                              h2::request<h2::empty_body>* req,
+                              h2::response<h2::string_body>* resp);
```
[the remark this answers is not in the document]
Comment (romange) [reply util/cloud/aws.h:39]: ok

File: util/cloud/s3.h [hunk 2]
```diff
@@ -18,7 +18,7 @@ using ListBucketsResult = io::Result<std::vector<std::string>>;
 // Empty if no more pages left.
 using ListObjectsResult = io::Result<std::string>;
 
-ListBucketsResult ListS3Buckets(const AWS& aws, http::Client* http_client);
+ListBucketsResult ListS3Buckets(AWS& aws, http::Client* http_client);
```
Comment (romange) [inline util/cloud/s3.h:21]: it could be this but I prefer to pass AWS by pointer (style guide thingy).

Comment (dranikpg) [reply util/cloud/s3.h:21]: > Also comment that we can be updated with a refreshed token in some cases during the operation run
Why here? Isn't this a property of all AWS operations now?

Comment (romange) [reply util/cloud/s3.h:21]: i mean, if you pass AWS by pointer, just explain what's the side effect of it. It's not clear from the signature why AWS is mutable,

File: examples/s3_demo.cc [hunk 3]
```diff
@@ -38,7 +38,7 @@ template <typename Body> std::ostream& operator<<(std::ostream& os, const h2::re
   return os;
 }
 
-void ListBuckets(const AWS& aws, ProactorBase* proactor) {
+void ListBuckets(AWS& aws, ProactorBase* proactor) {
```
Comment (romange) [inline examples/s3_demo.cc:41]: still need to pass AWS by pointer

File: util/cloud/s3.h [hunk 4]
```diff
@@ -18,7 +18,8 @@ using ListBucketsResult = io::Result<std::vector<std::string>>;
 // Empty if no more pages left.
 using ListObjectsResult = io::Result<std::string>;
 
-ListBucketsResult ListS3Buckets(const AWS& aws, http::Client* http_client);
+// List all S3 buckets. Refresh AWS token if needed.
+ListBucketsResult ListS3Buckets(AWS* aws, http::Client* http_client);
```
Comment (dranikpg) [inline util/cloud/s3.h:22]: Added

File: util/cloud/aws.cc [hunk 5]
```diff
@@ -403,25 +417,56 @@ void AWS::Sign(std::string_view payload_sig, HttpHeader* header) const {
...
+                                 h2::request<h2::empty_body>* req,
+                                 h2::response<h2::string_body>* resp) {
+  Sign(payload_sig, req);
+
+  VLOG(1) << "Req: " << *req;
+  auto ec = client->Send(*req, resp);
+
+  // Check if session_token expired on an aws connection, established via instance metadata.
+  // In this case, try to update it, re-sign the request and re-try it.
+  if (!ec && resp->result() == h2::status::bad_request && !connection_data_.role_name.empty() &&
```
Comment (romange) [inline util/cloud/aws.cc:430]: maybe just wrap 2 conditions (bad_request and find kExpiredTokenSentinel) in an auxillary function `IsTokenRefreshNeeded` or something

File: util/cloud/aws.cc [hunk 6]
```diff
@@ -403,25 +417,56 @@ void AWS::Sign(std::string_view payload_sig, HttpHeader* header) const {
...
+
+    auto updated_data = GetConnectionDataFromMetadata(connection_data_.role_name);
+    if (!updated_data)
+      return ec;
+
+    PopulateAwsConnectionData(std::move(*updated_data), &connection_data_);
+    SetScopeAndSignKey();
+
+    // Re-connect client if needed.
+    if ((*resp)["Connection"] == "close") {
```
Comment (romange) [inline util/cloud/aws.cc:442]: is it needed in your experience? i.e. do you get conn-close when a token is expired?

**Review by romange — COMMENTED** [review]
lgtm, minor comments

**Review by romange — APPROVED** [review]

**Pull request merged**